src/Elements/Geometry/Ray.cs

```diff
@@ -77,6 +78,102 @@ public bool Intersects(Triangle tri, out Vector3 result)
+        /// <param name="result">The intersection result.</param>
+        /// <returns>True if an intersection occurs, otherwise false. If true, check the intersection result for the location of the intersection.</returns>
+        internal bool Intersects(Solid solid, out Vector3 result)
+        {
+            var distanceToClosestResult = double.MaxValue;
+            var faces = solid.Faces;
+            var intersects = false;
+            var finalResult = default(Vector3);
+            foreach (var face in faces)
+            {
+                if (Intersects(face.Value, out Vector3 tempResult))
+                {
+                    intersects = true;
+                    var newDistance = Origin.DistanceTo(tempResult);
+                    if (newDistance < distanceToClosestResult)
+                    {
+                        distanceToClosestResult = newDistance;
+                        finalResult = tempResult;
+                    }
+                }
+            }
+            result = finalResult;
+            return intersects;
+        }
+
+        /// <summary>
+        /// Does this ray intersect with the provided face?
+        /// </summary>
+        /// <param name="face">The Face to intersect with.</param>
+        /// <param name="result">The intersection result.</param>
+        /// <returns>True if an intersection occurs, otherwise false. If true, check the intersection result for the location of the intersection.</returns>
+        internal bool Intersects(Face face, out Vector3 result)
+        {
+            var plane = face.Plane();
+            if (Intersects(plane, out Vector3 intersection))
+            {
+                var boundaryPolygon = face.Outer.ToPolygon();
+                var transformToPolygon = new Transform(plane.Origin, plane.Normal);
+                var transformFromPolygon = new Transform(transformToPolygon);
+                transformFromPolygon.Invert();
+                var transformedPolygon = transformFromPolygon.OfPolygon(boundaryPolygon);
+                var transformedIntersection = transformFromPolygon.OfVector(intersection);
+                if(transformedPolygon.Contains(transformedIntersection))
+                {
+                    result = intersection;
+                    return true;
+                }
+            }
+            result = default(Vector3);
+            return false;
+        }
+
+        /// <summary>
+        /// Does this ray intersect the provided plane?
+        /// </summary>
+        /// <param name="plane">The Plane to intersect with.</param>
+        /// <param name="result">The intersection result.</param>
+        /// <returns>True if an intersection occurs, otherwise false — this can occur if the ray is very close to parallel to the plane.
+        /// If true, check the intersection result for the location of the intersection.</returns>
+        public bool Intersects(Plane plane, out Vector3 result)
```

Comment:
This is not required for this PR, but might be nice...This code and the `Line.Intersects` code will share the same logic for intersection testing. The line intersection code simply checks if the intersection is past the end of the line as its final step. Can we use the same intersection logic in both places?

Reply:
Done!